File: snippets/security/input-validation.js
/**
 * SNIPPET: Input Validation Middleware
 * CATEGORY: Security
 * LANGUAGE: JavaScript (Express)
 * STATUS: ✅ Ready
 * 
 * DESCRIPTION: 
 *   Server-side input validation using express-validator.
 *   Prevents injection attacks and ensures data integrity.
 *   NEVER trust client-side validation alone!
 * 
 * DEPENDENCIES: 
 *   npm install express-validator
 * 
 * USAGE:
 *   const { validateUser, validateId } = require('./middleware/validation');
 *   
 *   app.post('/users', validateUser, (req, res) => { ... });
 *   app.get('/users/:id', validateId, (req, res) => { ... });
 * 
 * EXAMPLE:
 *   // Validation automatically returns 400 with error details
 *   // POST /users with invalid email returns:
 *   // { "errors": [{ "msg": "Invalid email", "path": "email" }] }
 */

const { body, param, query, validationResult } = require('express-validator');

// ============================================
// VALIDATION RESULT HANDLER
// ============================================

/**
 * Middleware to check validation results and return errors
 */
const handleValidationErrors = (req, res, next) => {
  const errors = validationResult(req);
  if (!errors.isEmpty()) {
    return res.status(400).json({ 
      error: 'Validation failed',
      errors: errors.array().map(err => ({
        field: err.path,
        message: err.msg
      }))
    });
  }
  next();
};

// ============================================
// COMMON VALIDATORS
// ============================================

/**
 * Validate user creation/update
 */
const validateUser = [
  body('email')
    .isEmail().withMessage('Invalid email address')
    .normalizeEmail(),
  
  body('name')
    .trim()
    .isLength({ min: 1, max: 100 }).withMessage('Name must be 1-100 characters')
    .escape(),  // Prevent XSS
  
  body('age')
    .optional()
    .isInt({ min: 0, max: 150 }).withMessage('Age must be 0-150'),
  
  handleValidationErrors
];

/**
 * Validate ID parameter (numeric)
 */
const validateId = [
  param('id')
    .isInt({ min: 1 }).withMessage('ID must be a positive integer'),
  
  handleValidationErrors
];

/**
 * Validate UUID parameter
 */
const validateUuid = [
  param('id')
    .isUUID().withMessage('Invalid UUID format'),
  
  handleValidationErrors
];

/**
 * Validate pagination query parameters
 */
const validatePagination = [
  query('page')
    .optional()
    .isInt({ min: 1 }).withMessage('Page must be positive integer')
    .toInt(),
  
  query('limit')
    .optional()
    .isInt({ min: 1, max: 100 }).withMessage('Limit must be 1-100')
    .toInt(),
  
  handleValidationErrors
];

/**
 * Validate search query
 */
const validateSearch = [
  query('q')
    .trim()
    .isLength({ min: 1, max: 200 }).withMessage('Search query must be 1-200 characters')
    .escape(),
  
  handleValidationErrors
];

// ============================================
// CUSTOM VALIDATORS
// ============================================

/**
 * Create custom field validator
 * @param {string} field - Field name
 * @param {object} options - Validation options
 */
const createFieldValidator = (field, options = {}) => {
  const validators = [];
  
  if (options.required) {
    validators.push(
      body(field).exists().withMessage(`${field} is required`)
    );
  }
  
  if (options.type === 'email') {
    validators.push(
      body(field).isEmail().withMessage('Invalid email').normalizeEmail()
    );
  }
  
  if (options.type === 'string') {
    validators.push(
      body(field)
        .trim()
        .isLength({ 
          min: options.minLength || 1, 
          max: options.maxLength || 255 
        })
        .withMessage(`${field} must be ${options.minLength || 1}-${options.maxLength || 255} characters`)
        .escape()
    );
  }
  
  if (options.type === 'number') {
    validators.push(
      body(field)
        .isNumeric().withMessage(`${field} must be a number`)
    );
  }
  
  return validators;
};

// ============================================
// SANITIZATION HELPERS
// ============================================

/**
 * Sanitize object by removing dangerous fields
 * @param {object} obj - Object to sanitize
 * @param {string[]} allowedFields - Whitelist of allowed field names
 */
const sanitizeObject = (obj, allowedFields) => {
  const sanitized = {};
  for (const field of allowedFields) {
    if (obj.hasOwnProperty(field)) {
      sanitized[field] = obj[field];
    }
  }
  return sanitized;
};

/**
 * Remove HTML tags from string
 * @param {string} str - String to sanitize
 */
const stripHtml = (str) => {
  if (typeof str !== 'string') return str;
  return str.replace(/<[^>]*>/g, '');
};

module.exports = {
  // Middleware
  handleValidationErrors,
  
  // Pre-built validators
  validateUser,
  validateId,
  validateUuid,
  validatePagination,
  validateSearch,
  
  // Helpers
  createFieldValidator,
  sanitizeObject,
  stripHtml,
  
  // Re-export express-validator for custom use
  body,
  param,
  query,
  validationResult
};

File: snippets/security/input-validation.py
"""
SNIPPET: Input Validation
CATEGORY: Security
LANGUAGE: Python (FastAPI)
STATUS: Ready

DESCRIPTION:
    Server-side input validation using Pydantic (built into FastAPI).
    Shows common validation patterns and sanitization helpers.

DEPENDENCIES:
    None (Pydantic is built into FastAPI)

USAGE:
    from input_validation import sanitize_string, CreateUserRequest

    @app.post("/users")
    async def create_user(body: CreateUserRequest):
        # body is already validated by Pydantic
        clean_name = sanitize_string(body.name)
        ...

IMPORTANT:
    - ALWAYS validate on the server, even if frontend validates too
    - NEVER trust client-side validation alone
    - Use Pydantic models for all request bodies
"""

import re
import html
from pydantic import BaseModel, Field, field_validator, EmailStr

# ---------------------------------------------------------------------------
# Example: Validated request models
# ---------------------------------------------------------------------------

class CreateUserRequest(BaseModel):
    """Example: validate a user creation request."""
    name: str = Field(min_length=1, max_length=100)
    email: EmailStr
    role: str = Field(default="user", pattern=r"^(user|expert|admin)$")
    age: int | None = Field(default=None, ge=0, le=150)

    @field_validator("name")
    @classmethod
    def clean_name(cls, v):
        # Strip HTML tags and excessive whitespace
        v = re.sub(r"<[^>]+>", "", v).strip()
        if not v:
            raise ValueError("Name cannot be empty after sanitization")
        return v


class SearchRequest(BaseModel):
    """Example: validate search/filter params."""
    query: str = Field(min_length=1, max_length=500)
    page: int = Field(default=1, ge=1)
    per_page: int = Field(default=20, ge=1, le=100)
    sort_by: str = Field(default="created_at", pattern=r"^(created_at|name|email|updated_at)$")
    sort_order: str = Field(default="desc", pattern=r"^(asc|desc)$")


class FileUploadRequest(BaseModel):
    """Example: validate file metadata."""
    filename: str = Field(max_length=255)
    content_type: str = Field(pattern=r"^(application|text|image)/[\w.+-]+$")
    size: int = Field(ge=1, le=50 * 1024 * 1024)  # 50MB max

    @field_validator("filename")
    @classmethod
    def safe_filename(cls, v):
        # Remove path traversal attempts
        v = v.replace("..", "").replace("/", "").replace("\\", "")
        if not v:
            raise ValueError("Invalid filename")
        return v


# ---------------------------------------------------------------------------
# Sanitization helpers
# ---------------------------------------------------------------------------

def sanitize_string(value: str) -> str:
    """Remove HTML tags, escape special chars, normalize whitespace."""
    # Remove HTML tags
    clean = re.sub(r"<[^>]+>", "", value)
    # Escape remaining HTML entities
    clean = html.escape(clean)
    # Normalize whitespace
    clean = re.sub(r"\s+", " ", clean).strip()
    return clean


def sanitize_url(url: str) -> str | None:
    """Validate and sanitize a URL. Returns None if invalid."""
    url = url.strip()
    if not re.match(r"^https?://", url, re.IGNORECASE):
        return None
    # Block javascript: URLs
    if re.match(r"^javascript:", url, re.IGNORECASE):
        return None
    return url


def sanitize_filename(filename: str) -> str:
    """Make a filename safe for storage."""
    # Remove path separators and traversal
    safe = re.sub(r'[/\\:*?"<>|]', "_", filename)
    safe = safe.replace("..", "_")
    return safe[:255]  # Limit length